snap/squashfs/stat.go

```diff
@@ -0,0 +1,330 @@
+	GID uint32
+}
+
+type stat struct {
+	path  string
+	size  int64
+	mode  os.FileMode
+	mtime time.Time
+	user  string
+	group string
+}
+
+func (s stat) Name() string       { return filepath.Base(s.path) }
+func (s stat) Size() int64        { return s.size }
+func (s stat) Mode() os.FileMode  { return s.mode }
+func (s stat) ModTime() time.Time { return s.mtime }
+func (s stat) IsDir() bool        { return s.mode.IsDir() }
+func (s stat) Sys() interface{}   { return nil }
+func (s stat) Path() string       { return s.path } // not path of os.FileInfo
+
+const minLen = 57 // "drwxrwxr-x user/user             53595 2017-12-08 11:19 ."
+
+func fromRaw(raw []byte) (*stat, error) {
+	if len(raw) < minLen {
+		return nil, errBadLine(raw)
+	}
+
+	st := &stat{}
+
+	// first, the file mode, e.g. "-rwxr-xr-x"; always that length (1+3*3==10)
+	if err := st.parseMode(raw[:10]); err != nil {
+		return nil, err
+	}
+
+	// next, user/group info
+	p := 10
+	if n, err := st.parseOwner(raw[p:]); err != nil {
+		return nil, err
+	} else {
+		p += n
+	}
+
+	// next'll come the size or the node type
+	if n, err := st.parseSize(raw[p:]); err != nil {
+		return nil, err
+	} else {
+		p += n
+	}
+
+	if err := st.parseTime(raw[p : p+16]); err != nil {
+		return nil, err
+	}
+
+	p += 16
+
+	if raw[p] != ' ' {
+		return nil, errBadLine(raw)
+	}
+	p++
+	if raw[p] != '.' {
+		return nil, errBadLine(raw)
+	}
+	p++
+	if len(raw[p:]) == 0 {
+		st.path = "/"
+	} else {
+		st.path = string(raw[p:])
+	}
+
+	return st, nil
+}
+
+type statError struct {
+	part string
+	raw  []byte
+}
+
+func (e statError) Error() string {
+	return fmt.Sprintf("cannot parse %s: %q", e.part, e.raw)
+}
+
+func errBadLine(raw []byte) statError {
+	return statError{
+		part: "line",
+		raw:  raw,
+	}
+}
+
+func errBadMode(raw []byte) statError {
+	return statError{
+		part: "mode",
+		raw:  raw,
+	}
+}
+
+func errBadOwner(raw []byte) statError {
+	return statError{
+		part: "owner",
+		raw:  raw,
+	}
+}
+
+func errBadNode(raw []byte) statError {
+	return statError{
+		part: "node",
+		raw:  raw,
+	}
+}
+
+func errBadSize(raw []byte) statError {
+	return statError{
+		part: "size",
+		raw:  raw,
+	}
+}
+
+func (st *stat) parseTime(raw []byte) error {
+	t, err := time.ParseInLocation("2006-01-02 15:04", string(raw), time.Local)
+	if err != nil {
+		return err
+	}
+
+	st.mtime = t
+
+	return nil
+}
+
+func (st *stat) parseMode(raw []byte) error {
+	switch raw[0] {
+	case '-':
+		// 0
+	case 'd':
+		st.mode |= os.ModeDir
+	case 's':
+		st.mode |= os.ModeSocket
+	case 'c':
+		st.mode |= os.ModeCharDevice
+	case 'b':
+		st.mode |= os.ModeDevice
+	case 'p':
+		st.mode |= os.ModeNamedPipe
+	case 'l':
+		st.mode |= os.ModeSymlink
+	default:
+		return errBadMode(raw)
+	}
+
+	for i := 0; i < 3; i++ {
+		m, me := modeFromTriplet(raw[1+3*i:4+3*i], uint(2-i))
+		if me != nil {
+			return me
+		}
+		st.mode |= m
+	}
+
+	return nil
+}
+
+func (st *stat) parseOwner(raw []byte) (int, error) {
+	var p, ui, uj, gi, gj int
```

Comment:
I personally would probably have choose longer var names here like `{user,group}{Start.End}Offset`- but less typing this way :)